File: Structural/Proxy/Proxy.go
package main

import "fmt"

type IDatbase interface {
	getUsers() string
}

type Database struct{}

var cache map[string]string

func newDatabase() IDatbase {
	return &Database{}
}

func (d *Database) getUsers() string {
	return "Users from database"
}

type CachedDb struct {
	db IDatbase
}

func newCachedDB(db IDatbase) IDatbase {
	return &CachedDb{
		db: newDatabase(),
	}
}

func (cd *CachedDb) getUsers() string {
	data, ok := cache["users"]
	if ok {
		return data
	}

	cache["users"] = "users"
	return "Found uses on cache"

}

type DBManager struct{}

func newDBManager() IDatbase {
	return &DBManager{}
}

func (dbm DBManager) getUsers() string {
	d := newDatabase()
	c := newCachedDB(d)

	return c.getUsers()
}

func main() {
	db := newDBManager()
	fmt.Println(db.getUsers())
	fmt.Println(db.getUsers())
}

File: Structural/Proxy/Proxy.py
from __future__ import annotations

from abc import ABC, abstractmethod


class IDatabase(ABC):
    @abstractmethod
    def get_users(self):
        pass

    @abstractmethod
    def get_user_info(self):
        pass


class DatabaseAccessor(IDatabase):

    def get_users(self):
        return ['kushal', 'nischal', 'ram', 'shyam']

    def get_user_info(self):
        return f"User info"


class CacheDbData(IDatabase):
    _cache = {}

    def get_users(self):
        if self._cache.get("users"):
            return f"Users found on cache"
        self._cache["users"] = "new Users"

        return self._cache

    def get_user_info(self):
        if self._cache.get("user_info"):
            return f"User info found in cache"

        self._cache['user_info'] = "user info"

        return self._cache


class DBManager(IDatabase):

    def __init__(self):
        self.service = CacheDbData()

    def get_users(self):
        return self.service.get_users()

    def get_user_info(self):
        return self.service.get_user_info()


data = DBManager()
print(data.get_users())
print(data.get_users())
